app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/CategoryProductHelper.java

```diff
@@ -0,0 +1,140 @@
+        categoryText.setText(bold(baseFragment.getString(R.string.txtCategories)));
+        categoryText.setMovementMethod(LinkMovementMethod.getInstance());
+        categoryText.append(" ");
+        categoryText.setClickable(true);
+        categoryText.setMovementMethod(LinkMovementMethod.getInstance());
+
+        if (categories.isEmpty()) {
+            categoryText.setVisibility(View.GONE);
+        } else {
+            categoryText.setVisibility(View.VISIBLE);
+            // Add all the categories to text view and link them to wikidata is possible
+            for (int i = 0, lastIndex = categories.size() - 1; i <= lastIndex; i++) {
+                CategoryName category = categories.get(i);
+                CharSequence categoryName = getCategoriesTag(category);
+                // Add category name to text view
+                categoryText.append(categoryName);
+                // Add a comma if not the last item
+                if (i != lastIndex) {
+                    categoryText.append(", ");
+                }
+
+                if(category.getCategoryTag().equals(baseFragment.getString(R.string.categorytag_en_alcoholic_beverages))){
+                    containsAlcohol = true;
+                }
+            }
+        }
+    }
+
+    private CharSequence getCategoriesTag(CategoryName category) {
+        SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder();
+        ClickableSpan clickableSpan = new ClickableSpan() {
+            @Override
+            public void onClick(View view) {
+                if (category.getIsWikiDataIdPresent()) {
+                    apiClient.doSomeThing(category.getWikiDataId(), (value, result) -> {
+                        if (value) {
+                            FragmentActivity activity = baseFragment.getActivity();
+
+                            if (activity != null && !activity.isFinishing()) {
+                                BottomScreenCommon.showBottomScreen(result, category,
+                                    activity.getSupportFragmentManager());
+                            }
+                        } else {
+                            ProductBrowsingListActivity.startActivity(baseFragment.getContext(),
+                                category.getCategoryTag(),
+                                category.getName(),
+                                SearchType.CATEGORY);
+                        }
+                    });
+                } else {
+                    ProductBrowsingListActivity.startActivity(baseFragment.getContext(),
+                        category.getCategoryTag(),
+                        category.getName(),
+                        SearchType.CATEGORY);
+                }
+            }
+        };
+        spannableStringBuilder.append(category.getName());
+        spannableStringBuilder.setSpan(clickableSpan, 0, spannableStringBuilder.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
+        if (!category.isNotNull()) {
```

Comment:
@harshlele I don't understand this line. Could you explain why is there a double negation (! and isNotNull) and also why this null check?

Reply:
i didnt write that method, just moved it to the new file because it was related to the categories